server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/variables/VariableDefinitionUtil.java

```diff
@@ -62,9 +69,80 @@
   public static final String REDEFINE_IMMEDIATELY_FOLLOW = "semantics.redefineImmediatelyFollow";
   public static final String LEVELS_MUST_MATCH = "semantics.levelsMustMatch";
   public static final String REDEFINED_CONTAIN_VALUE = "semantics.redefinedContainValue";
+  private static final String PIC_AND_USAGE_CONFLICT = "semantics.picAndUsageConflict";
 
   public static final ErrorSeverity SEVERITY = ERROR;
 
+  static final List<UsageFormat> BLANK_AND_SIGN_NOT_ALLOWED_USAGE =
+          ImmutableList.of(UTF_8, NATIONAL);
+
+
+  static final ImmutableList<UsageFormat> STRING_USAGE_FORMAT =
+          ImmutableList.of(NATIONAL, DISPLAY_1, UTF_8, DISPLAY);
+  static final ImmutableList<UsageFormat> NUMERIC_DATATYPE_USAGE_FORMAT =
+          ImmutableList.of(
+                  COMP_1,
+                  COMPUTATIONAL_1,
+                  COMP_2,
+                  COMPUTATIONAL_2,
+                  COMP,
+                  COMPUTATIONAL,
+                  COMPUTATIONAL_3,
+                  COMP_3,
+                  COMP_4,
+                  COMPUTATIONAL_4,
+                  COMP_5,
+                  COMPUTATIONAL_5,
+                  NATIONAL, DISPLAY);
+  static final ResultWithErrors<Boolean> FALSE_RESULT_WITH_NO_ERRORS =
+          new ResultWithErrors<>(false, ImmutableList.of());
+  private static final Function<ElementaryNodes, ResultWithErrors<Boolean>> CHECK_STRING_USAGE =
+          variableNode -> {
+            boolean isError =
+                    !STRING_USAGE_FORMAT.contains(variableNode.getUsageFormat())
+                            && variableNode.getUsageFormat() != UNDEFINED;
+            if (isError)
+              return new ResultWithErrors<>(
+                      true,
+                      ImmutableList.of(
+                              variableNode.getError(
+                                      MessageTemplate.of(PIC_AND_USAGE_CONFLICT), ErrorSeverity.WARNING)));
+            return FALSE_RESULT_WITH_NO_ERRORS;
+          };
+  private static final List<EffectiveDataType> NUMERIC_DATA_TYPES = ImmutableList.of(REAL, INTEGER);
+  private static final Function<ElementaryNodes, ResultWithErrors<Boolean>> CHECK_NUMERIC_USAGE =
+          variableNode -> {
+            if (NUMERIC_DATATYPE_USAGE_FORMAT.contains(variableNode.getUsageFormat())) {
+              boolean isError = !NUMERIC_DATA_TYPES.contains(variableNode.getEffectiveDataType());
+              if (isError)
```

Comment:
Seems these conditions may be collapsed, at least if you extract the conditional part to a method

Reply:
Done